File: src/setCoaddEdgeBits.cc
// -*- LSST-C++ -*-

/* 
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 * 
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the LSST License Statement and 
 * the GNU General Public License along with this program.  If not, 
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */
 
/**
* @file
*
* @author Russell Owen
*/
#include "boost/format.hpp"

#include "lsst/pex/exceptions.h"
#include "lsst/coadd/utils/setCoaddEdgeBits.h"

namespace pexExcept = lsst::pex::exceptions;
namespace afwImage = lsst::afw::image;
namespace coaddUtils = lsst::coadd::utils;

/**
* @brief set edge bits of coadd mask based on weight map
*
* @throw pexExcept::InvalidParameterException if the dimensions of coaddMask and weightMap do not match.
*/
template<typename WeightPixelT>
void coaddUtils::setCoaddEdgeBits(
    // spell out lsst:afw::image to make Doxygen happy
    lsst::afw::image::Mask<lsst::afw::image::MaskPixel> &coaddMask, ///< mask of coadd
    lsst::afw::image::Image<WeightPixelT> const &weightMap  ///< weight map
) {
    typedef afwImage::Mask<afwImage::MaskPixel>::x_iterator MaskXIter;
    typedef typename afwImage::Image<WeightPixelT>::const_x_iterator WeightMapConstXIter;

    if (coaddMask.getDimensions() != weightMap.getDimensions()) {
        throw LSST_EXCEPT(pexExcept::InvalidParameterException,
            (boost::format("coaddMask and weightMap dimensions differ: %dx%d != %dx%d") %
            coaddMask.getWidth() % coaddMask.getHeight() % weightMap.getWidth() % weightMap.getHeight()
            ).str());
    }
    
    afwImage::MaskPixel const edgeMask = afwImage::Mask<afwImage::MaskPixel>::getPlaneBitMask("EDGE");

    // Set the pixels row by row, to avoid repeated checks for end-of-row
    for (int y = 0, endY = weightMap.getHeight(); y != endY; ++y) {
        WeightMapConstXIter weightMapPtr = weightMap.row_begin(y);
        WeightMapConstXIter const weightMapEndPtr = weightMap.row_end(y);
        MaskXIter coaddMaskPtr = coaddMask.row_begin(y);
        for (; weightMapPtr != weightMapEndPtr; ++weightMapPtr, ++coaddMaskPtr) {
            if (*weightMapPtr == 0) {
                (*coaddMaskPtr) = (*coaddMaskPtr) | edgeMask;
            }
        }
    }
}

//
// Explicit instantiations
//
#define INSTANTIATE(WEIGHTPIXEL) \
    template void coaddUtils::setCoaddEdgeBits<WEIGHTPIXEL>( \
        afwImage::Mask<afwImage::MaskPixel> &coaddMask, \
        afwImage::Image<WEIGHTPIXEL> const &weightMap \
    );

INSTANTIATE(double);
INSTANTIATE(float);
INSTANTIATE(int);
INSTANTIATE(boost::uint16_t);

File: tests/setCoaddEdgeBits.py
#!/usr/bin/env python

# 
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
# 
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the LSST License Statement and 
# the GNU General Public License along with this program.  If not, 
# see <http://www.lsstcorp.org/LegalNotices/>.
#

"""Test lsst.coadd.utils.setCoaddEdgeBits
"""
import os
import math
import pdb # we may want to say pdb.set_trace()
import unittest

import numpy

import lsst.afw.image as afwImage
import lsst.afw.image.testUtils as imTestUtils
import lsst.utils.tests as utilsTests
import lsst.pex.logging as pexLog
import lsst.coadd.utils as coaddUtils

Verbosity = 0 # increase to see trace
pexLog.Trace_setVerbosity("lsst.coadd.utils", Verbosity)

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-


class SetCoaddEdgeBitsTestCase(unittest.TestCase):
    """A test case for setCoaddEdgeBits
    """
    def testRandomMap(self):
        """Test setCoaddEdgeBits using a random depth map
        """
        imShape = (50, 50)
        coaddMask = afwImage.MaskU(imShape[0], imShape[1], 0)
        

        numpy.random.seed(12345)
        depthMapArray = numpy.random.randint(0, 3, list(imShape))
        depthMap = imTestUtils.imageFromArray(depthMapArray, afwImage.ImageU)
        
        refCoaddMaskArray = imTestUtils.arrayFromMask(coaddMask)
        edgeMask = afwImage.MaskU.getPlaneBitMask("EDGE")
        refCoaddMaskArray |= numpy.where(depthMapArray > 0, 0, edgeMask)
        
        coaddUtils.setCoaddEdgeBits(coaddMask, depthMap)
        coaddMaskArray = imTestUtils.arrayFromMask(coaddMask)
        if numpy.any(refCoaddMaskArray != coaddMaskArray):
            errMsgList = (
                "Coadd mask does not match reference=%s:" % (badPixelMask,),
                "computed=  %s" % (coaddMaskArray,),
                "reference= %s" % (refCoaddMaskArray,),
            )
            errMsg = "\n".join(errMsgList)
            self.fail(errMsg)

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Return a suite containing all the test cases in this module.
    """
    utilsTests.init()

    suites = []
    suites += unittest.makeSuite(SetCoaddEdgeBitsTestCase)
    suites += unittest.makeSuite(utilsTests.MemoryTestCase)

    return unittest.TestSuite(suites)

if __name__ == "__main__":
    utilsTests.run(suite())
